src/net/discovery.rs

```diff
@@ -0,0 +1,172 @@
+//! A discovery service to gather addressing info collected from gossip Join and ForwardJoin messages.
+
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    sync::{Arc, RwLock},
+    time::Duration,
+};
+
+use iroh::discovery::{Discovery, DiscoveryError, DiscoveryItem, NodeData, NodeInfo};
+use iroh_base::NodeId;
+use n0_future::{
+    boxed::BoxStream,
+    stream::{self, StreamExt},
+    task::AbortOnDropHandle,
+    time::SystemTime,
+};
+
+pub(crate) struct RetentionOpts {
+    /// How long to keep received node info records alive before pruning them
+    retention: Duration,
+    /// How often to check for expired entries
+    evict_interval: Duration,
+}
+
+impl Default for RetentionOpts {
+    fn default() -> Self {
+        Self {
+            retention: Duration::from_secs(60 * 5),
+            evict_interval: Duration::from_secs(30),
+        }
+    }
+}
+
+/// A static node discovery that expires nodes after some time.
+///
+/// It is added to the endpoint when constructing a gossip instance, and the gossip actor
+/// then adds node addresses as received with Join or ForwardJoin messages.
+#[derive(Debug, Clone)]
+pub(crate) struct GossipDiscovery {
+    nodes: NodeMap,
+    _task_handle: Arc<AbortOnDropHandle<()>>,
+}
+
+type NodeMap = Arc<RwLock<BTreeMap<NodeId, StoredNodeInfo>>>;
+
+#[derive(Debug)]
+struct StoredNodeInfo {
+    data: NodeData,
+    last_updated: SystemTime,
+}
+
+impl Default for GossipDiscovery {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GossipDiscovery {
+    const PROVENANCE: &'static str = "gossip";
+
+    /// Creates a new gossip discovery instance.
+    pub(crate) fn new() -> Self {
+        Self::with_opts(Default::default())
+    }
+
+    pub(crate) fn with_opts(opts: RetentionOpts) -> Self {
+        let nodes: NodeMap = Default::default();
+        let task = {
+            let nodes = Arc::downgrade(&nodes);
+            n0_future::task::spawn(async move {
+                loop {
+                    n0_future::time::sleep(opts.evict_interval).await;
```

Comment:
I think it is better to use an interval, instead of sleep?

Reply:
fixed